nvme_gw_persistence.py

```diff
@@ -0,0 +1,404 @@
+            with rados.WriteOpCtx() as write_op:
+                # Set exclusive parameter to fail write_op if object exists
+                write_op.new(rados.LIBRADOS_CREATE_EXCLUSIVE)
+                self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                    (str(self.version),))
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+                self.logger.info(
+                    f"First gateway: created object {self.omap_name}")
+        except rados.ObjectExists:
+            self.logger.info(f"{self.omap_name} omap object already exists.")
+
+    def _write_key(self, key: str, val: str):
+        """Writes key and value to the persistent config."""
+
+        try:
+            version_update = int(self.version) + 1
+            with rados.WriteOpCtx() as write_op:
+                # Compare operation failure will cause write failure
+                write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
+                                  rados.LIBRADOS_CMPXATTR_OP_EQ)
+                self.ioctx.set_omap(write_op, (key,), (val,))
+                self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                    (str(version_update),))
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+            self.version = version_update
+            self.logger.debug(f"omap_key generated: {key}")
+        except Exception as ex:
+            self.logger.error(f"Unable to write to omap: {ex}. Exiting!")
+            raise
+
+    def _delete_key(self, key: str):
+        """Deletes key from omap persistent config."""
+
+        version_update = int(self.version) + 1
+        with rados.WriteOpCtx() as write_op:
+            # Compare operation failure will cause delete failure
+            write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
+                              rados.LIBRADOS_CMPXATTR_OP_EQ)
+            self.ioctx.remove_omap_keys(write_op, (key,))
+            self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                (str(version_update),))
+            self.ioctx.operate_write_op(write_op, self.omap_name)
+        self.version = version_update
+        self.logger.debug(f"omap_key deleted: {key}")
+
+    def add_bdev(self, bdev_name: str, val: str):
+        """Adds a bdev to the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._write_key(key, val)
+
+    def delete_bdev(self, bdev_name: str):
+        """Deletes a bdev from the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._delete_key(key)
+
+    def _restore_bdevs(self, omap_dict, callback):
+        """Restores a bdev from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.BDEV_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                req = pb2.bdev_create_req(
```

Comment:
nit: I would prefer we would just pass `args` to the callbacks (or explicitly specified every arg as a callback param) and build requests (`pb2.bdev_create_req`)  inside the callback. This way the persistent config would not need to know about `pb2` at all. It looks cleaner to me but I don't insist.